warehouse/manage/views.py

```diff
@@ -1439,65 +1444,154 @@ def manage_project_roles(project, request, _form_class=CreateRoleForm):
                 queue="error",
             )
         else:
-            request.db.add(
-                Role(user=user, project=project, role_name=form.role_name.data)
-            )
-            request.db.add(
-                JournalEntry(
-                    name=project.name,
-                    action=f"add {role_name} {username}",
-                    submitted_by=request.user,
-                    submitted_from=request.remote_addr,
+            token_service = request.find_service(ITokenService, name="email")
+            token = token_service.dumps(
+                {
+                    "action": "email-project-role-verify",
+                    "desired_role": role_name,
+                    "user_id": user.id,
+                    "project_id": project.id,
+                    "submitter_id": request.user.id,
+                }
+            )
+            user_invite = (
+                request.db.query(RoleInvitation)
+                .filter(RoleInvitation.user == user)
+                .filter(RoleInvitation.project == project)
+                .one_or_none()
+            )
+
+            send_email = False
+            if user_invite:
+                if (
+                    user_invite.invite_status == RoleInvitationStatus.Pending.value
+                    or user_invite.invite_status == RoleInvitationStatus.Accepted.value
+                ):
+                    request.session.flash(
+                        f"User '{username}' already has an active invite/role. "
+                        f"Please try again later.",
+                        queue="error",
+                    )
+                else:
+                    send_email = True
+                    user_invite.invite_status = RoleInvitationStatus.Pending.value
+                    user_invite.token = token
+            else:
+                send_email = True
+                request.db.add(
+                    RoleInvitation(
+                        user=user,
+                        project=project,
+                        invite_status=RoleInvitationStatus.Pending.value,
+                        token=token,
+                    )
                 )
```

Comment:
Can we refactor this so the `send_email` boolean isn't necessary, and so that the conditionals are only one deep? We can always return early if we need to.